proxy/src/metrics.rs

```diff
@@ -1,156 +1,348 @@
+    pub connection_failures_breakdown: CounterVec<ConnectionFailuresBreakdownSet>,
+    /// Number of bytes sent/received between all clients and backends.
+    pub io_bytes: CounterVec<StaticLabelSet<Direction>>,
+    /// Number of errors by a given classification.
+    pub errors_total: CounterVec<StaticLabelSet<crate::error::ErrorKind>>,
+}
 
-pub static CONSOLE_REQUEST_LATENCY: Lazy<HistogramVec> = Lazy::new(|| {
-    register_histogram_vec!(
-        "proxy_console_request_latency",
-        "Time it took for proxy to establish a connection to the compute endpoint",
-        // proxy_wake_compute/proxy_get_role_info
-        &["request"],
-        // largest bucket = 2^16 * 0.2ms = 13s
-        exponential_buckets(0.0002, 2.0, 16).unwrap(),
-    )
-    .unwrap()
-});
+impl ProxyMetrics {
+    pub fn new() -> Self {
+        Self {
+            db_connections: NumDbConnectionsGauge::default(),
+            client_connections: NumClientConnectionsGauge::default(),
+            connection_requests: NumConnectionRequestsGauge::default(),
+            http_endpoint_pools: HttpEndpointPools::default(),
+            compute_connection_latency_seconds: measured::HistogramVec::new_metric_vec(
+                ComputeConnectionLatencySet::default(),
+                // largest bucket = 2^16 * 0.5ms = 32s
+                Thresholds::exponential_buckets(0.0005, 2.0),
+            ),
+            compute_console_request_latency: measured::HistogramVec::new_metric_vec(
+                ConsoleRequestSet {
+                    request: ThreadedRodeo::with_hasher(BuildHasherDefault::default()),
+                },
+                // largest bucket = 2^16 * 0.2ms = 13s
+                Thresholds::exponential_buckets(0.0002, 2.0),
+            ),
+            // largest bucket = 2^16 * 0.2ms = 13s
+            control_plane_token_acquire_seconds: measured::Histogram::new_metric(
+                Thresholds::exponential_buckets(0.0002, 2.0),
+            ),
+            // largest bucket = 3^16 * 0.05ms = 2.15s
+            http_conn_content_length_bytes: measured::Histogram::new_metric(
+                Thresholds::exponential_buckets(8.0, 2.0),
+            ),
+            // 1us -> 65ms
+            http_pool_reclaimation_lag_seconds: measured::Histogram::new_metric(
+                Thresholds::exponential_buckets(1e-6, 2.0),
+            ),
+            http_pool_opened_connections: Gauge::new(),
+            allowed_ips_cache_misses: CounterVec::default(),
+            allowed_ips_number: measured::Histogram::new_metric(Thresholds::with_buckets([
+                0.0, 1.0, 2.0, 3.0, 4.0, 5.0, 10.0, 20.0, 50.0, 100.0,
+            ])),
+            accepted_connections_by_sni: CounterVec::default(),
+            connection_failures_total: CounterVec::default(),
+            connection_failures_breakdown: CounterVec::default(),
+            io_bytes: CounterVec::default(),
+            errors_total: CounterVec::default(),
+        }
+    }
+}
```

Comment:
this is quite noisy. it would be nice to be able to generate this with the MetricGroup derive

Reply:
done in measured 0.0.13